runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -375,22 +391,47 @@ impl<T: Config> Pallet<T> {
 				&mut rng,
 			);
 
-		// Assure the maximum block weight is adhered, by limiting bitfields and backed
-		// candidates. Dispute statement sets were already limited before.
-		let non_disputes_weight = apply_weight_limit::<T>(
-			&mut backed_candidates,
-			&mut bitfields,
-			max_block_weight.saturating_sub(checked_disputes_sets_consumed_weight),
-			&mut rng,
-		);
+		let full_weight = if context == ProcessInherentDataContext::ProvideInherent {
+			// Assure the maximum block weight is adhered, by limiting bitfields and backed
+			// candidates. Dispute statement sets were already limited before.
+			let non_disputes_weight = apply_weight_limit::<T>(
+				&mut backed_candidates,
+				&mut bitfields,
+				max_block_weight.saturating_sub(checked_disputes_sets_consumed_weight),
+				&mut rng,
+			);
 
-		let full_weight = non_disputes_weight.saturating_add(checked_disputes_sets_consumed_weight);
+			let full_weight =
+				non_disputes_weight.saturating_add(checked_disputes_sets_consumed_weight);
 
-		METRICS.on_after_filter(full_weight.ref_time());
+			METRICS.on_after_filter(full_weight.ref_time());
 
-		if full_weight.any_gt(max_block_weight) {
-			log::warn!(target: LOG_TARGET, "Post weight limiting weight is still too large.");
-		}
+			if full_weight.any_gt(max_block_weight) {
+				log::warn!(target: LOG_TARGET, "Post weight limiting weight is still too large.");
+			}
+
+			full_weight
+		} else {
```

Comment:
I think, this would be more readable if moved to the top, right after where we get the weights. We also don't need if/else - just

```rust
 if let ProcessInherentDataContext::Enter = context {
.... panic 
}
```

Apart from that I am not convinced we are solving any real issue here, but if we want to avoid overweight blocks in all cases, I would prefer to just include the above filtering in the weights. This should be fine with our deterministic dispute import, because the node should not provide redundant data in the first place anymore. Anyhow I thought about this and did not consider this a real issue, hence I kept it as is.

The reason, why I would prefer this over the panic is that it re-introduces possible inconsistencies, I intended to get rid of. If for any reason those two code paths arrive at (slightly) different weights, the we could end up DoSing ourselves.

Given that we are calculating the weight in very different manners in those code paths, this does not sound like a too unrealistic threat. For now we should be overestimating in `ProvideInherent`, so it should be fine, but if we optimized this at some point, things might change.

Reply:
Update: Just accounting for the weight, does not actually fix the issue. So disregard that part.